In [5]:
%pip install pandas

   ---------------------------------------- 0.0/10.0 MB ? eta -:--:--
   ------------------------ --------------- 6.0/10.0 MB 46.3 MB/s eta 0:00:01
   ---------------------------------------- 10.0/10.0 MB 50.6 MB/s  0:00:00
   ---------------------------------------- 0.0/12.6 MB ? eta -:--:--
   ---------------------------------------  12.6/12.6 MB 72.9 MB/s eta 0:00:01
   ---------------------------------------- 12.6/12.6 MB 50.3 MB/s  0:00:00

   ---------------------------------------- 0/3 [tzdata]
   ---------------------------------------- 0/3 [tzdata]
   ---------------------------------------- 0/3 [tzdata]
   ---------------------------------------- 0/3 [tzdata]
   ------------- -------------------------- 1/3 [numpy]
   ------------- -------------------------- 1/3 [numpy]
   ------------- -------------------------- 1/3 [numpy]
   ------------- -------------------------- 1/3 [numpy]
   ------------- -------------------------- 1/3 [numpy]
   ------------- -------------------------

  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.

[notice] A new release of pip is available: 26.1.2 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [6]:
import pandas as pd
import os 

In [17]:
sales = pd.read_csv(r"c:\users\dionr\Desktop\E-Commerce-Sales-Customer-Analytics\data\raw\Sales.csv")
sales.head()

,Order Number,Line Item,Order Date,Delivery Date,CustomerKey,StoreKey,ProductKey,Quantity,Currency Code
0,366000,1,1/1/2016,NaN,265598,10,1304,1,CAD
1,366001,1,1/1/2016,1/13/2016,1269051,0,1048,2,USD
2,366001,2,1/1/2016,1/13/2016,1269051,0,2007,1,USD
3,366002,1,1/1/2016,1/12/2016,266019,0,1106,7,CAD
4,366002,2,1/1/2016,1/12/2016,266019,0,373,1,CAD


ROW AND COLUMNS

In [18]:
sales.shape

(62884, 9)

In [19]:
sales.info()

<class 'pandas.DataFrame'>
RangeIndex: 62884 entries, 0 to 62883
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   Order Number   62884 non-null  int64
 1   Line Item      62884 non-null  int64
 2   Order Date     62884 non-null  str  
 3   Delivery Date  13165 non-null  str  
 4   CustomerKey    62884 non-null  int64
 5   StoreKey       62884 non-null  int64
 6   ProductKey     62884 non-null  int64
 7   Quantity       62884 non-null  int64
 8   Currency Code  62884 non-null  str  
dtypes: int64(6), str(3)
memory usage: 4.3 MB


In [20]:
sales.columns

Index(['Order Number', 'Line Item', 'Order Date', 'Delivery Date',
       'CustomerKey', 'StoreKey', 'ProductKey', 'Quantity', 'Currency Code'],
      dtype='str')

NULL MISSING VALUES

In [21]:
sales.isnull().sum()

Order Number         0
Line Item            0
Order Date           0
Delivery Date    49719
CustomerKey          0
StoreKey             0
ProductKey           0
Quantity             0
Currency Code        0
dtype: int64

In [23]:
missing = sales.isnull().sum()

missing_percentage = (missing / len(sales)) * 100

pd.DataFrame({
    "Missing Values": missing,
    "Missing Percentage": missing_percentage.round(2)
})

,Missing Values,Missing Percentage
Order Number,0,0.00
Line Item,0,0.00
Order Date,0,0.00
Delivery Date,49719,79.06
CustomerKey,0,0.00
StoreKey,0,0.00
ProductKey,0,0.00
Quantity,0,0.00
Currency Code,0,0.00


CHECK DUPLICATES

In [24]:
sales.duplicated().sum()

np.int64(0)

In [28]:
sales[sales.duplicated()].head()

,Order Number,Line Item,Order Date,Delivery Date,CustomerKey,StoreKey,ProductKey,Quantity,Currency Code


CHECK NUMERICAL COLUMNS

In [26]:
sales.describe()

,Order Number,Line Item,CustomerKey,StoreKey,ProductKey,Quantity
count,6.288400e+04,62884.000000,6.288400e+04,62884.000000,62884.000000,62884.000000
mean,1.430905e+06,2.164207,1.180797e+06,31.802144,1125.859344,3.144790
std,4.532963e+05,1.365170,5.859634e+05,22.978188,709.244010,2.256371
min,3.660000e+05,1.000000,3.010000e+02,0.000000,1.000000,1.000000
25%,1.121017e+06,1.000000,6.808580e+05,8.000000,437.000000,1.000000
50%,1.498016e+06,2.000000,1.261200e+06,37.000000,1358.000000,2.000000
75%,1.788010e+06,3.000000,1.686496e+06,53.000000,1650.000000,4.000000
max,2.243032e+06,7.000000,2.099937e+06,66.000000,2517.000000,10.000000


CHECK DATA RANGE

In [29]:
sales["Order Date"].min(), sales["Order Date"].max()

('1/1/2016', '9/9/2020')

CHECK CURRENCIES

In [31]:
sales["Currency Code"].value_counts()

Currency Code
USD    33767
EUR    12621
GBP     8140
CAD     5415
AUD     2941
Name: count, dtype: int64

Initial Data Quality Findings

The initial profiling of the Sales dataset identified several areas requiring further investigation:

1. Order Date and Delivery Date are stored as strings rather than datetime values.
2. Delivery Date contains 49,719 missing values (approximately 79.1% of records).
3. No exact duplicate rows were identified.
4. StoreKey contains a value of 0, which requires investigation.
5. Sales transactions contain five currencies: USD, EUR, GBP, CAD, and AUD.
6. Quantity ranges from 1 to 10 with no immediately apparent invalid negative quantities.
7. The dataset covers sales from January 1, 2016 to September 9, 2020.

Further profiling will be performed before any data-cleaning decisions are made.

In [32]:
sales[sales["StoreKey"] == 0].head()

,Order Number,Line Item,Order Date,Delivery Date,CustomerKey,StoreKey,ProductKey,Quantity,Currency Code
1,366001,1,1/1/2016,1/13/2016,1269051,0,1048,2,USD
2,366001,2,1/1/2016,1/13/2016,1269051,0,2007,1,USD
3,366002,1,1/1/2016,1/12/2016,266019,0,1106,7,CAD
4,366002,2,1/1/2016,1/12/2016,266019,0,373,1,CAD
5,366002,3,1/1/2016,1/12/2016,266019,0,1080,4,CAD


In [33]:
sales["StoreKey"].value_counts().sort_index().head(10)

StoreKey
0     13165
1       292
2        18
4       431
5       892
6       615
8      1360
9      1577
10     1269
12      224
Name: count, dtype: int64

In [34]:
sales[sales["StoreKey"] == 0].shape

(13165, 9)

INSPECTING STORES

In [35]:
stores = pd.read_csv("../data/raw/Stores.csv")

In [36]:
stores.info()

<class 'pandas.DataFrame'>
RangeIndex: 67 entries, 0 to 66
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   StoreKey       67 non-null     int64  
 1   Country        67 non-null     str    
 2   State          67 non-null     str    
 3   Square Meters  66 non-null     float64
 4   Open Date      67 non-null     str    
dtypes: float64(1), int64(1), str(3)
memory usage: 2.7 KB


In [37]:
stores.head()

,StoreKey,Country,State,Square Meters,Open Date
0,1,Australia,Australian Capital Territory,595.0,1/1/2008
1,2,Australia,Northern Territory,665.0,1/12/2008
2,3,Australia,South Australia,2000.0,1/7/2012
3,4,Australia,Tasmania,2000.0,1/1/2010
4,5,Australia,Victoria,2000.0,12/9/2015


In [38]:
stores.shape

(67, 5)

In [39]:
stores["StoreKey"].min(), stores["StoreKey"].max()

(np.int64(0), np.int64(66))

In [40]:
stores["StoreKey"].value_counts().sort_index().head(10)

StoreKey
0    1
1    1
2    1
3    1
4    1
5    1
6    1
7    1
8    1
9    1
Name: count, dtype: int64

STOREKEY THAT HAS 0 VALUE = MISSING DELIVERY DATA

In [42]:
sales.groupby("StoreKey")["Delivery Date"].apply(
    lambda x: x.isna().sum()
).sort_values(ascending=False).head(10)

StoreKey
9     1577
50    1519
55    1518
54    1498
61    1485
59    1472
45    1471
57    1442
44    1436
65    1395
Name: Delivery Date, dtype: int64

In [43]:
pd.crosstab(
    sales["StoreKey"] == 0,
    sales["Delivery Date"].isna()
)

Delivery Date,False,True
StoreKey,,
False,0,49719
True,13165,0


In [44]:
stores[stores["StoreKey"] == 0]

,StoreKey,Country,State,Square Meters,Open Date
66,0,Online,Online,NaN,1/1/2010


In [45]:
store_sales = (
    sales.groupby("StoreKey")
    .size()
    .reset_index(name="Transactions")
    .merge(stores, on="StoreKey", how="left")
    .sort_values("Transactions", ascending=False)
)

store_sales.head(15)

,StoreKey,Transactions,Country,State,Square Meters,Open Date
0,0,13165,Online,Online,NaN,1/1/2010
7,9,1577,Canada,Northwest Territories,1500.0,3/4/2005
44,50,1519,United States,Kansas,2000.0,3/6/2008
48,55,1518,United States,Nevada,2000.0,12/15/2009
47,54,1498,United States,Nebraska,2000.0,6/7/2013
52,61,1485,United States,South Carolina,2000.0,12/15/2012
51,59,1472,United States,Oregon,2000.0,8/8/2012
40,45,1471,United States,Connecticut,2000.0,7/8/2007
50,57,1442,United States,New Mexico,1645.0,6/3/2010
39,44,1436,United States,Arkansas,2000.0,6/3/2010


In [46]:
sales[sales["Delivery Date"].isna()]["Currency Code"].value_counts()

Currency Code
USD    26555
EUR    10067
GBP     6643
CAD     4206
AUD     2248
Name: count, dtype: int64

In [47]:
pd.crosstab(
    sales["Currency Code"],
    sales["Delivery Date"].isna(),
    normalize="index"
).round(3)

Delivery Date,False,True
Currency Code,,
AUD,0.236,0.764
CAD,0.223,0.777
EUR,0.202,0.798
GBP,0.184,0.816
USD,0.214,0.786


In [ ]:
sales.groupby("StoreKey")["Delivery Date"].apply(
    lambda x: x.isna().sum()
).sort_values(ascending=False).head(10)

Delivery Date Investigation

The Sales dataset contains 49,719 missing Delivery Date values.

Initial investigation showed that the missing values are not randomly distributed. StoreKey = 0 represents the Online sales channel, and all 13,165 online transactions have a Delivery Date.

A cross-tabulation showed:

- StoreKey = 0 (Online): 13,165 transactions with Delivery Date
- StoreKey > 0 (Physical stores): 49,719 transactions without Delivery Date

Therefore, the missing Delivery Date values appear to be structural rather than erroneous. Delivery dates are relevant to online transactions, while physical-store purchases do not require delivery.

The missing values should therefore not be automatically filled or removed during data cleaning.

In [48]:
delivery_by_channel = pd.crosstab(
    sales["StoreKey"] == 0,
    sales["Delivery Date"].isna(),
    normalize="index"
).round(4) * 100

delivery_by_channel

Delivery Date,False,True
StoreKey,,
False,0.0,100.0
True,100.0,0.0


In [49]:
currency_channel = pd.crosstab(
    sales["Currency Code"],
    sales["StoreKey"] == 0,
    normalize="index"
).round(4) * 100

currency_channel

StoreKey,False,True
Currency Code,,
AUD,76.44,23.56
CAD,77.67,22.33
EUR,79.76,20.24
GBP,81.61,18.39
USD,78.64,21.36


In [60]:
file_path = r"C:\Users\dionr\Desktop\E-Commerce-Sales-Customer-Analytics\data\raw\Customers.csv"

for enc in ("utf-8", "cp1252", "latin1"):
    try:
        customers = pd.read_csv(file_path, encoding=enc, low_memory=False)
        break
    except UnicodeDecodeError:
        continue
else:
    # fallback using python engine if needed
    customers = pd.read_csv(file_path, encoding="latin1", engine="python", low_memory=False)

customers.head()

,CustomerKey,Gender,Name,City,State Code,State,Zip Code,Country,Continent,Birthday
0,301,Female,Lilly Harding,WANDEARAH EAST,SA,South Australia,5523,Australia,Australia,7/3/1939
1,325,Female,Madison Hull,MOUNT BUDD,WA,Western Australia,6522,Australia,Australia,9/27/1979
2,554,Female,Claire Ferres,WINJALLOK,VIC,Victoria,3380,Australia,Australia,5/26/1947
3,786,Male,Jai Poltpalingada,MIDDLE RIVER,SA,South Australia,5223,Australia,Australia,9/17/1957
4,1042,Male,Aidan Pankhurst,TAWONGA SOUTH,VIC,Victoria,3698,Australia,Australia,11/19/1965


In [61]:
customers.info()

<class 'pandas.DataFrame'>
RangeIndex: 15266 entries, 0 to 15265
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   CustomerKey  15266 non-null  int64
 1   Gender       15266 non-null  str  
 2   Name         15266 non-null  str  
 3   City         15266 non-null  str  
 4   State Code   15256 non-null  str  
 5   State        15266 non-null  str  
 6   Zip Code     15266 non-null  str  
 7   Country      15266 non-null  str  
 8   Continent    15266 non-null  str  
 9   Birthday     15266 non-null  str  
dtypes: int64(1), str(9)
memory usage: 1.2 MB


In [62]:
customers.shape

(15266, 10)

In [63]:
customers.head()

,CustomerKey,Gender,Name,City,State Code,State,Zip Code,Country,Continent,Birthday
0,301,Female,Lilly Harding,WANDEARAH EAST,SA,South Australia,5523,Australia,Australia,7/3/1939
1,325,Female,Madison Hull,MOUNT BUDD,WA,Western Australia,6522,Australia,Australia,9/27/1979
2,554,Female,Claire Ferres,WINJALLOK,VIC,Victoria,3380,Australia,Australia,5/26/1947
3,786,Male,Jai Poltpalingada,MIDDLE RIVER,SA,South Australia,5223,Australia,Australia,9/17/1957
4,1042,Male,Aidan Pankhurst,TAWONGA SOUTH,VIC,Victoria,3698,Australia,Australia,11/19/1965


In [64]:
customers.isnull().sum()

CustomerKey     0
Gender          0
Name            0
City            0
State Code     10
State           0
Zip Code        0
Country         0
Continent       0
Birthday        0
dtype: int64

In [65]:
customers.duplicated().sum()

np.int64(0)

In [66]:
customers.describe()

,CustomerKey
count,1.526600e+04
mean,1.060508e+06
std,6.127097e+05
min,3.010000e+02
25%,5.140335e+05
50%,1.079244e+06
75%,1.593980e+06
max,2.099937e+06


In [67]:
customers["CustomerKey"].nunique(), len(customers)

(15266, 15266)

In [68]:
customers[customers["State Code"].isna()]

,CustomerKey,Gender,Name,City,State Code,State,Zip Code,Country,Continent,Birthday
5304,729681,Female,Rossana Padovesi,Polvica,NaN,Napoli,80035,Italy,Europe,4/18/1981
5316,732289,Male,Indro Piccio,Varcaturo,NaN,Napoli,80014,Italy,Europe,2/24/1949
5372,742042,Male,Amaranto Loggia,Casaferro,NaN,Napoli,80034,Italy,Europe,3/14/1936
5377,742886,Female,Edmonda Capon,Terzigno,NaN,Napoli,80040,Italy,Europe,8/6/1963
5378,743343,Female,Ambra Sagese,Pomigliano D'Arco,NaN,Napoli,80038,Italy,Europe,1/5/1961
5485,759705,Male,Callisto Lo Duca,Casilli,NaN,Napoli,80047,Italy,Europe,8/28/1976
5525,765589,Male,Michelino Lucchesi,Pompei Scavi,NaN,Napoli,80045,Italy,Europe,11/13/1947
5531,766410,Male,Adelmio Beneventi,Licola,NaN,Napoli,80078,Italy,Europe,1/13/1940
5631,781667,Female,Ilda Manna,Napoli,NaN,Napoli,80134,Italy,Europe,5/8/1977
5695,789177,Male,Calogero Folliero,Mariglianella,NaN,Napoli,80030,Italy,Europe,3/3/2000


In [69]:
customers[customers["State Code"].isna()][
    ["CustomerKey", "City", "State", "Country", "Zip Code"]
]

,CustomerKey,City,State,Country,Zip Code
5304,729681,Polvica,Napoli,Italy,80035
5316,732289,Varcaturo,Napoli,Italy,80014
5372,742042,Casaferro,Napoli,Italy,80034
5377,742886,Terzigno,Napoli,Italy,80040
5378,743343,Pomigliano D'Arco,Napoli,Italy,80038
5485,759705,Casilli,Napoli,Italy,80047
5525,765589,Pompei Scavi,Napoli,Italy,80045
5531,766410,Licola,Napoli,Italy,80078
5631,781667,Napoli,Napoli,Italy,80134
5695,789177,Mariglianella,Napoli,Italy,80030


In [70]:
sales_customer_keys = set(sales["CustomerKey"])
customer_keys = set(customers["CustomerKey"])

invalid_customer_keys = sales_customer_keys - customer_keys

len(invalid_customer_keys)

0

In [71]:
sales["CustomerKey"].nunique()

11887

In [72]:
customers_with_sales = sales["CustomerKey"].nunique()

customers_total = customers["CustomerKey"].nunique()

customers_without_sales = customers_total - customers_with_sales

customers_with_sales, customers_without_sales

(11887, 3379)

Customer Data Quality Findings

The Customers dataset contains 15,266 unique customers across 10 attributes.

Initial profiling identified:

1. CustomerKey is unique across all 15,266 customer records and can serve as the primary key.
2. No completely duplicated customer records were found.
3. Only 10 State Code values are missing, representing approximately 0.07% of customers.
4. Birthday is stored as a string and should be converted to a datetime type during data cleaning.
5. Zip Code is stored as a string, which is appropriate because postal codes are identifiers rather than numerical measurements.
6. Referential integrity between Sales.CustomerKey and Customers.CustomerKey will be validated before proceeding with data cleaning.

In [78]:
customers_with_sales_keys = sales["CustomerKey"].unique()

customers_without_sales_df = customers[
    ~customers["CustomerKey"].isin(customers_with_sales_keys)
]

customers_without_sales_df.head()

,CustomerKey,Gender,Name,City,State Code,State,Zip Code,Country,Continent,Birthday
3,786,Male,Jai Poltpalingada,MIDDLE RIVER,SA,South Australia,5223,Australia,Australia,9/17/1957
5,1086,Male,Hayden Clegg,TEMPLERS,SA,South Australia,5371,Australia,Australia,1/20/1954
6,1133,Male,Nicholas Caffyn,JUBILEE POCKET,QLD,Queensland,4802,Australia,Australia,11/22/1969
7,1256,Male,Lincoln Jenks,KULLOGUM,QLD,Queensland,4660,Australia,Australia,3/12/1950
13,1677,Male,Dean Innes,ELLESMERE,QLD,Queensland,4610,Australia,Australia,11/22/1964


In [79]:
customers_without_sales_df["Country"].value_counts()

Country
United States     1122
Australia          640
Canada             374
United Kingdom     374
Germany            323
France             232
Netherlands        199
Italy              115
Name: count, dtype: int64

In [80]:
customers_without_sales_df["Continent"].value_counts()

Continent
North America    1496
Europe           1243
Australia         640
Name: count, dtype: int64

In [81]:
customers_without_sales_df["Gender"].value_counts()

Gender
Male      1719
Female    1660
Name: count, dtype: int64

In [82]:
customers[customers["State"] == "Napoli"]["State Code"].value_counts(dropna=False)

State Code
NaN    10
Name: count, dtype: int64

Customer Investigation Findings

Further investigation identified 3,379 customers with no corresponding sales transactions. These customers are distributed across all major markets, with the largest groups located in the United States, Australia, Canada, and the United Kingdom. They will not be classified as inactive customers without additional evidence.

The 10 missing State Code values all belong to customers whose State is listed as Napoli, Italy. No other State Code values are associated with Napoli in the dataset, so the missing values cannot be reliably inferred from the available data. They will therefore be retained as NULL rather than replaced with potentially incorrect values.

CustomerKey referential integrity was also validated: every CustomerKey in the Sales table exists in the Customers table.

In [87]:
products = pd.read_csv("../data/raw/Products.csv")

In [88]:
products.info()

<class 'pandas.DataFrame'>
RangeIndex: 2517 entries, 0 to 2516
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   ProductKey      2517 non-null   int64
 1   Product Name    2517 non-null   str  
 2   Brand           2517 non-null   str  
 3   Color           2517 non-null   str  
 4   Unit Cost USD   2517 non-null   str  
 5   Unit Price USD  2517 non-null   str  
 6   SubcategoryKey  2517 non-null   int64
 7   Subcategory     2517 non-null   str  
 8   CategoryKey     2517 non-null   int64
 9   Category        2517 non-null   str  
dtypes: int64(3), str(7)
memory usage: 196.8 KB


In [89]:
products.shape

(2517, 10)

In [90]:
products.head()

,ProductKey,Product Name,Brand,Color,Unit Cost USD,Unit Price USD,SubcategoryKey,Subcategory,CategoryKey,Category
0,1,Contoso 512MB MP3 Player E51 Silver,Contoso,Silver,$6.62,$12.99,101,MP4&MP3,1,Audio
1,2,Contoso 512MB MP3 Player E51 Blue,Contoso,Blue,$6.62,$12.99,101,MP4&MP3,1,Audio
2,3,Contoso 1G MP3 Player E100 White,Contoso,White,$7.40,$14.52,101,MP4&MP3,1,Audio
3,4,Contoso 2G MP3 Player E200 Silver,Contoso,Silver,$11.00,$21.57,101,MP4&MP3,1,Audio
4,5,Contoso 2G MP3 Player E200 Red,Contoso,Red,$11.00,$21.57,101,MP4&MP3,1,Audio


In [91]:
products.isnull().sum()

ProductKey        0
Product Name      0
Brand             0
Color             0
Unit Cost USD     0
Unit Price USD    0
SubcategoryKey    0
Subcategory       0
CategoryKey       0
Category          0
dtype: int64

In [92]:
products.duplicated().sum()

np.int64(0)

In [93]:
products.describe()

,ProductKey,SubcategoryKey,CategoryKey
count,2517.000000,2517.000000,2517.000000
mean,1259.000000,491.810091,4.878824
std,726.739637,229.887134,2.299170
min,1.000000,101.000000,1.000000
25%,630.000000,305.000000,3.000000
50%,1259.000000,406.000000,4.000000
75%,1888.000000,801.000000,8.000000
max,2517.000000,808.000000,8.000000


In [94]:
products["ProductKey"].nunique(), len(products)

(2517, 2517)

In [95]:
products["Brand"].value_counts()

Brand
Contoso                 710
Fabrikam                267
Litware                 264
Proseware               244
Adventure Works         192
Southridge Video        192
Wide World Importers    173
The Phone Company       152
Tailspin Toys           144
A. Datum                132
Northwind Traders        47
Name: count, dtype: int64

In [96]:
products["Category"].value_counts()

Category
Home Appliances                  661
Computers                        606
Cameras and camcorders           372
Cell phones                      285
TV and Video                     222
Games and Toys                   166
Audio                            115
Music, Movies and Audio Books     90
Name: count, dtype: int64

In [97]:
products[["Unit Cost USD", "Unit Price USD"]].head(10)

,Unit Cost USD,Unit Price USD
0,$6.62,$12.99
1,$6.62,$12.99
2,$7.40,$14.52
3,$11.00,$21.57
4,$11.00,$21.57
5,$11.00,$21.57
6,$11.00,$21.57
7,$30.58,$59.99
8,$30.58,$59.99
9,$30.58,$59.99


In [108]:
sales_product_keys = set(sales["ProductKey"])
product_keys = set(products["ProductKey"])

invalid_product_keys = sales_product_keys - product_keys

len(invalid_product_keys)

0

In [109]:
sales["ProductKey"].nunique()

2492

In [110]:
products_sold = sales["ProductKey"].nunique()
products_total = products["ProductKey"].nunique()

products_not_sold = products_total - products_sold

products_sold, products_not_sold

(2492, 25)

Product Data Quality Findings

The Products dataset contains 2,517 products across 10 attributes.

Initial profiling identified:

1. ProductKey is unique across all 2,517 records and can serve as the primary key.
2. No missing values were identified.
3. No completely duplicated product records were identified.
4. Unit Cost USD and Unit Price USD are stored as strings containing currency symbols and must be converted to numeric values before financial calculations.
5. Products are organized into a Category → Subcategory → Product hierarchy.
6. The catalog contains 11 brands and 8 product categories.
7. Product assortment size should not be interpreted as sales performance; sales volume, revenue, and profitability will be analyzed separately.
8. Referential integrity between Sales.ProductKey and Products.ProductKey will be validated before cleaning.

In [114]:
stores = pd.read_csv("../data/raw/Stores.csv")

In [115]:
stores.head()

,StoreKey,Country,State,Square Meters,Open Date
0,1,Australia,Australian Capital Territory,595.0,1/1/2008
1,2,Australia,Northern Territory,665.0,1/12/2008
2,3,Australia,South Australia,2000.0,1/7/2012
3,4,Australia,Tasmania,2000.0,1/1/2010
4,5,Australia,Victoria,2000.0,12/9/2015


In [116]:
stores.isnull().sum()

StoreKey         0
Country          0
State            0
Square Meters    1
Open Date        0
dtype: int64

In [117]:
stores.duplicated().sum()

np.int64(0)

In [118]:
stores["StoreKey"].nunique(), len(stores)

(67, 67)

In [119]:
stores["Country"].value_counts()

Country
United States     24
Germany            9
France             7
United Kingdom     7
Australia          6
Canada             5
Netherlands        5
Italy              3
Online             1
Name: count, dtype: int64

In [120]:
stores["State"].nunique()

67

In [121]:
stores["Open Date"].min(), stores["Open Date"].max()

('1/1/2008', '8/8/2012')

In [122]:
stores[stores["Square Meters"].isna()]

,StoreKey,Country,State,Square Meters,Open Date
66,0,Online,Online,NaN,1/1/2010
